### DISCUSSION TITLE
Start-up / shut-down tasks?

### DISCUSSION THREAD

**Discussion by mindplay-dk** [lorenzofox3/zora#104] [Q&A] [answered]
Is there any way to perform some tasks before the test-suite starts, and after it ends?

(I don't mean before/after each test, but before/after the whole test-suite.)

I'm launching an `express` server for some functional tests - I'd like to keep this running for the duration of the entire test-suite, and then need to dispose of the server at the end.

I managed to "emulate" a start-up task this way:

```ts
import express from "express";
import type { Server } from "http";

let testServer: Promise<Server>; // 👈 singleton

export async function getTestServer(): Promise<Server> {
  return testServer || (testServer = createTestServer()); // 👈 initialize singleton
}

async function createTestServer(): Promise<Server> {
  return new Promise(resolve => {
    const app = express();

    // ... configure express...

    const server = app.listen(options.port, () => resolve(server));
  });
}
```

But how do I shut down?

I need to somehow wait for zora's test-runner finish, so I know when to shut down the test-server - for as long as it's up, the application doesn't end, which means the test hangs after it finishes.

I could start it up and shut it down for every individual test, but that will slow things down a lot - and it's a stateless server, so there's no practical reason to do that. I'd like to reuse it across all tests.

(incidentally, this is a good example of why I argued [here](https://github.com/lorenzofox3/zora/issues/101#issuecomment-922739762) that being required to explicitly kick off the test-runner would be "a good thing" - an obvious guess would have been something `await run()` after queing up your tests...)

Is there still a way to manually create your test-harness? That bit appears to have vanished from the README with version 5. The code of course is still there, but doesn't seem to have anything for this.

I could add a decorator around the reporter, I guess - but that seems a bit hacky and probably not really what that's for?

**Comment by mindplay-dk** [comment, marked as answer]
My solution for now is, I take control of the test-harness myself - I added a `test.ts`:

```ts
import { createHarness } from "zora";

export const harness = createHarness({});

export const { test } = harness;
```

Calling `createHarness` disables the default, global harness.

In all of my tests, I `import { test } from "./harness"` instead of `from "zora"`.

In the test entry-point script, I can then manually `await` the harness:

```ts
import { harness } from "./harness";
import { createJSONReporter } from "zora";
import { getTestServer } from "./test-server"; // 👆 see above

(async () => {

  await harness.report({
    reporter: createJSONReporter(),
  });

  const server = await getTestServer();

  await new Promise(resolve => server.close(resolve));
  
})();
```

Gets the job done. 👍

As discussed [here](https://github.com/lorenzofox3/zora/issues/101#issuecomment-922739762)) though, this honestly still leaves me wishing that the default harness didn't even exist - global state, and things executing without your starting them, is pretty confusing, and it seems unnecessary. Creating the harness and starting the reporter saves you literally two lines of code.

Just my opinion, but I'd really like to see `zora` be a pure *library* - the automatic starting and reporting makes it feel more like a *framework*, and it seems like we already have `pta` filling that role.

**Reply by lorenzofox3** [reply]
you got it right. Thanks for sharing your opinion but I simply disagree with the effects on the dev experience ;)

**Reply by mindplay-dk** [reply]
yeah, I can only speak for myself - but two lines of boilerplate to get started would have saved me 1-2 hours of reading the source-code.

It's really great code, by the way! Nice, simple and elegant. The only thing I disagree with is the use of global state and side-effects - that part is very confusing, with `createHarness` suppressing the default harness, the option to enable/disable global side-effects on the exported `test` etc. harness properties... you could get rid of this complexity and "mystery" and all it would cost is two lines of boilerplate.

I know you've said, it's a main goal, "[you don't need a test-runner](https://github.com/lorenzofox3/zora/issues/101#issuecomment-923920986)" - but the thing is, you do need a test-runner, whether that reality is visible or hidden from you.

For what it's worth, your own perspective is not the normal dev experience - since you built it, you know precisely what to expect. I dunno, maybe I'm the only one having problems with this, heh. 😊

Would there be some way to have a nested export maybe? So `import { createHarness } from "zora/core"` or something - that way, you wouldn't need to track any global state, or elicit any side-effects, etc... If someone wants the global harness and `test` etc, they would import it from `zora` - if they prefer to write their own front-end, they can import from `zora/core`, which wouldn't have any global state or side-effects.

Would that work?

**Reply by lorenzofox3** [reply]
technically yes, but That would mean to ship again two versions (cjs required as well here) of pretty much the same code 🤔. There is not much added value either as you are still able to use your own harness with the current set up. Maybe we can try something with rollup multiple entry points set up

**Reply by lorenzofox3** [reply]
https://github.com/lorenzofox3/zora/pull/106 if you want to play around

**Reply by mindplay-dk** [reply]
> There is not much added value either as you are still able to use your own harness with the current set up.

Yeah, the point of this idea was just to get rid of the "undo" code and side-effects in the internal codebase - making it easier to understand for someone reading the source-code... so, if you import from `zora`, you get the default, global harness - and if you import from `zora/core`, you get no side-effects, no global state and no default harness, like a pure library.

This way, global state and side-effects is a *static* opt-in, as opposed to a *dynamic* opt-out.

Make sense? :-)

**Reply by lorenzofox3** [reply]
sure. My points are: 
* side effects cause no dammage so far.  (the trouble would start if you happen to load multiple versions of zora at the same time introducing multiple "singletons")
* they are not of your business (as consumer). You would not even had noticed without looking at the code.

so based on that: is it worth it to add complexity ? (costs vs gain basically)

**Reply by lorenzofox3** [reply]
So. I won't invest much time on the "issue" at the moment. You can play around with the linked branch it you want to come up with an simple solution

**Reply by mindplay-dk** [reply]
> side effects [...] are not of your business (as consumer). You would not even had noticed without looking at the code.

Where you notice is where your test-suite runs without your starting it - that's the immediate surprise.

My test is a program that calls out to libraries - other libraries generally don't start running things without my calling something. That's essentially the difference between a library and a framework, right? Frameworks run and let you tag along via hooks (in this case `test`) - whereas libraries sit idle until you tell them to run.

> is it worth it to add complexity ? (costs vs gain basically)

The point of my raising this issue was to *remove* complexity, not adding it. 🙂

I can't recall if I already posted this?

https://github.com/lorenzofox3/zora/compare/master...mindplay-dk:no-undo

Just a quick draft, removing the global state, the "undo" code, the side-effects, the `onlyMode` option and the `hold` function - it's removals only, this doesn't add anything new.

This way, if you want a "framework", it's `zora`, and if you want a pure library, that's `zora/core` - if you use the pure library, the global harness is never even created.

(It's completely untested! I don't know if I'm using `exports` correctly - just a draft to show the idea.)

**Reply by lorenzofox3** [reply]
> This way, if you want a "framework", it's zora, and if you want a pure library, that's zora/core - if you use the pure library, the global harness is never even created.

Yes, I like this idea too. You'll have to make sure you don't break the current behavior though (for people calling ``import ... from 'zora'``. I'll get a look at the PR

**Reply by mindplay-dk** [reply]
I need to get better at `npm` and this package stuff - I *should* be able to actually test and PR this change... Just usually too busy coding 😅

**Reply by mindplay-dk** [reply]
I finally got the test-suite running (duh, rtfm) and my changes don't seem to break anything, so something isn't covered by tests - this should definitely be a breaking change in a couple of places...

I added `exports` to the root `package.json`, and maybe that's wrong? I don't fully understand how this `workspaces` feature works - the `package.json` in the root, and the one the `zora` folder both have `"name": "zora"`, so that's pretty confusing.

Which one is actually the package? I'm guessing the one in the `zora` folder, so I probably did it wrong. I don't understand yet what the `package.json` in the root is for then - is that just for the "development project" and not for any of the packages that actually get published?

(any idea where I can learn this stuff? the official documentation just seems to assume you already know what everything is... aiii...)

**Reply by mindplay-dk** [reply]
Okay, I get it now - the root package isn't really a "package" per se. It never gets published - it's just a project that provides a structure (workspace) for the actual packages. (I learned from [this video](https://www.youtube.com/watch?v=c9dB86KNuDU) and reading the [documentation](https://docs.npmjs.com/cli/v7/using-npm/workspaces).)

So I definitely got it wrong: the root package shouldn't have `exports` at all - it never gets published, so it can't export anything.

With that in mind, I corrected both `package.json` files, and configured rollup to produce separate modules: https://github.com/mindplay-dk/zora/commit/fe65288e4dd322422c43cf4fbd44436961413aa7

This resulted in chunking (modules with ugly hashes in the names) for two reasons:

1. Both `index` and `harness` import from `test`, and rollup doesn't like to duplicate code.

2. Both `index` and `harness` import from `zora-reporters`, which gets bundled into the `zora` package.

I went ahead and configured rollup to simply emit all the input modules one to one: https://github.com/mindplay-dk/zora/commit/e3a8ef859784dcfb2b5d0121e3d6a6510123db38

That also means `zora-reporters` is now treated as `external`, meaning it no longer gets bundled - otherwise, because it's a dependency in two places, we would get chunks.

I don't know if this was done intentionally? It isn't defined as either a dependency or a dev-dependency, so it doesn't get installed for the `zora` package, and I don't fully understand how you can load it in the first place? Maybe this is part of how workspaces work - enabling you to load packages you haven't actually listed as dependencies? In the tutorial video (above) they were adding their dependencies.

I'm unsure, but probably the dependency should be added, now that `zora-reporters` is treated as `external`: https://github.com/mindplay-dk/zora/commit/055462f2f70bc4baf3d2a9b44d20cb72d9587c30

I think this makes sense? Since `zora-reporters` gets published as a separate module, it shouldn't also be bundled internally with `zora`, sine that would make it impossible to update one or the other independently. It also creates a weird situation, if you actually install `zora-reporters` as a dependency in your project, since the default harness would reference the *bundled* version of `zora-reporters`, while a custom harness would reference *your installed* version, right?

I put that last change on a separate branch, in case I'm wrong.

I'm kind of fumbling around here. 😅

*Footnote:* one thing that threw me off was, both the root package and the actual `zora` package have `"name": "zora"` in `package.json` - that happens to work because the root package never gets installed as a dependency, but it's maybe a little confusing? I would suggest maybe `zora-workspace` for the root package. I don't think it strictly needs a name at all? Since it never gets published. But it's not the `zora` package you get on npm, so it might be nice to disambiguate?

**Reply by lorenzofox3** [reply]
> I finally got the test-suite running (duh, rtfm) and my changes don't seem to break anything, so something isn't covered by tests - this should definitely be a breaking change in a couple of places...

Yes: I had not finished to migrate some of the tests (the ones with the extension ``.todo``). The ones dealing with ``only mode`` should break in your case.

> I added `exports` to the root `package.json`, and maybe that's wrong? I don't fully understand how this `workspaces` feature works - the `package.json` in the root, and the one the `zora` folder both have `"name": "zora"`, so that's pretty confusing.

The root package is not a package per se (not published). The name does not really matter and I should simply remove it. It simply helps with the scaffolding of various projects within the same git repository (cf npm workspaces).

> Which one is actually the package? I'm guessing the one in the `zora` folder, so I probably did it wrong. I don't understand yet what the `package.json` in the root is for then - is that just for the "development project" and not for any of the packages that actually get published?

Yes you are correct. You can google for npm workspaces for more info.

The pacakges published are: zora, pta and zora-reporters

**Reply by lorenzofox3** [reply]
> Okay, I get it now - the root package isn't really a "package" per se. It never gets published - it's just a project that provides a structure (workspace) for the actual packages. (I learned from [this video](https://www.youtube.com/watch?v=c9dB86KNuDU) and reading the [documentation](https://docs.npmjs.com/cli/v7/using-npm/workspaces).)

correct.

> So I definitely got it wrong: the root package shouldn't have `exports` at all - it never gets published, so it can't export anything.

correct again. Only there for development

> With that in mind, I corrected both `package.json` files, and configured rollup to produce separate modules: [mindplay-dk@fe65288](https://github.com/mindplay-dk/zora/commit/fe65288e4dd322422c43cf4fbd44436961413aa7)

That is the idea. Although I would create a _core_ file which exports not only the harness but also the ``Assert`` prototype and eventually some other functions (to create protocol messages for example)

> This resulted in chunking (modules with ugly hashes in the names) for two reasons:
> 
>     1. Both `index` and `harness` import from `test`, and rollup doesn't like to duplicate code.
> 
>     2. Both `index` and `harness` import from `zora-reporters`, which gets bundled into the `zora` package.
> 
> 
> I went ahead and configured rollup to simply emit all the input modules one to one: [mindplay-dk@e3a8ef8](https://github.com/mindplay-dk/zora/commit/e3a8ef859784dcfb2b5d0121e3d6a6510123db38)

That is the idea: you can create multiple entry points (core and index in our case) and rollup should then be able to factorize common chunks into "utility files" which avoid code duplication. (you can have a look at rollup doc if you want more info)

> That also means `zora-reporters` is now treated as `external`, meaning it no longer gets bundled - otherwise, because it's a dependency in two places, we would get chunks.

That's problematic because zora needs to be bundled together with the two default reporters (TAP and JSON) which are platform agnostic.

> I don't know if this was done intentionally? It isn't defined as either a dependency or a dev-dependency, so it doesn't get installed for the `zora` package, and I don't fully understand how you can load it in the first place? Maybe this is part of how workspaces work - enabling you to load packages you haven't actually listed as dependencies? In the tutorial video (above) they were adding their dependencies.

Yes everything is intentional. You don't need to (and should not) list zora-reporters as a dependency: node is able to resolve it at development time thanks to the workspaces set up. 
When Rollup builds the artefacts for zora, it is also able to resolve the code it needs from zora-reporters (and trim all what is not relevant). So goes the same with the ``assertion`` workspace which is defined as an independent project (but do not get published)
So then, I only need to publish the built artefacts to npm and that's it. For the consumer, it is transparent: it gets the artefacts defined by zora's package.json and does not have to know the development set up and internal dependency tree. Once published zora is dependency free and fully bundled in one file (or two because I need the CJS version for legacy project and TS, enventually)

``zora-reporters`` get published too on its side to get the bin file and all the reporters (not only the ones zora needs) but technically has nothing to do with zora itself.

> I'm unsure, but probably the dependency should be added, now that `zora-reporters` is treated as `external`: [mindplay-dk@055462f](https://github.com/mindplay-dk/zora/commit/055462f2f70bc4baf3d2a9b44d20cb72d9587c30)
> 
> I think this makes sense? Since `zora-reporters` gets published as a separate module, it shouldn't also be bundled internally with `zora`, sine that would make it impossible to update one or the other independently. It also creates a weird situation, if you actually install `zora-reporters` as a dependency in your project, since the default harness would reference the _bundled_ version of `zora-reporters`, while a custom harness would reference _your installed_ version, right?

You should not see zora-reporters as a dependency of zora. The bundled version of zora happens to have the code related to the TAP and  the JSON reporter but the fact it comes from the ``zora-reporters`` workspace is just a _coincidence_ and an organisational trick. 

The zora-reporters pacakge comes as an independent package with a lot more (the diff reporter, the cli and any new reporter). If I put  the TAP and the JSON reporter into the zora-reporters workspace, it is just a matter of organization.

> I put that last change on a separate branch, in case I'm wrong.
> 
> I'm kind of fumbling around here. 😅
> 
> _Footnote:_ one thing that threw me off was, both the root package and the actual `zora` package have `"name": "zora"` in `package.json` - that happens to work because the root package never gets installed as a dependency, but it's maybe a little confusing? I would suggest maybe `zora-workspace` for the root package. I don't think it strictly needs a name at all? Since it never gets published. But it's not the `zora` package you get on npm, so it might be nice to disambiguate?

yes the name in the root package is not useful at all and I should simply remove it

**Reply by mindplay-dk** [reply]
> you can create multiple entry points (core and index in our case) and rollup should then be able to factorize common chunks into "utility files" which avoid code duplication

I wasn't sure if you would be okay with that - some developers are opposed to publishing bundler artifacts for node packages. I'm not really for or against myself - but if you're going to do that, it should probably ship with source-maps, which it also isn't set up for right now?

> > That also means zora-reporters is now treated as external, meaning it no longer gets bundled [...]
>
> That's problematic because zora needs to be bundled together with the two default reporters (TAP and JSON) which are platform agnostic.
> [...]
>The zora-reporters pacakge comes as an independent package with a lot more (the diff reporter, the cli and any new reporter). If I put the TAP and the JSON reporter into the zora-reporters workspace, it is just a matter of organization.

I would suggest you move the TAP and JSON reporters to the core, then?

Since they have to be bundled anyhow, they're not really an optional dependency - they are a part of the core functionality. It might feel good to put them in `zora-reporters` because they're reporters, but as you said, it's just a matter of organization, and arguably this is core functionality, since the core doesn't function without it.

That also avoids having the bundler produce artifacts - probably don't need source-maps either then.

And it removes any concerns about `zora-reporters` having more dependencies.

**Reply by lorenzofox3** [reply]
> I wasn't sure if you would be okay with that - some developers are opposed to publishing bundler artifacts for node packages. I'm not really for or against myself - but if you're going to do that, it should probably ship with source-maps, which it also isn't set up for right now?

You don't need source map here as there is no "source" in the published package, see the [file served by the CDN for example](https://unpkg.com/zora@5.0.0/dist/index.js). The bundler simply concatenates files, there is no compilation, obfuscation etc.
Again, the rest are details of the development flow. From the consumer perspective, zora is juste one file of source code, that's it.

> I would suggest you move the TAP and JSON reporters to the core, then?
>
>Since they have to be bundled anyhow, they're not really an optional dependency - they are a part of the core functionality. It might feel good to put them in zora-reporters because they're reporters, but as you said, it's just a matter of organization, and arguably this is core functionality, since the core doesn't function without it.

No thanks you, from a development perspective, I prefer to find all the reporters in the ``reporters`` directory

**Reply by mindplay-dk** [reply]
Just for the record, you understand the bundler is currently *copying* from zora-reporters to the published files, right?

It wouldn't matter if `reporters` was just an internal package/repository (implementation detail) - the only reason I bring it up is because `zora-reporters` is also being published as a separate package, which means you end up with two copies (two different versions) of the two bundled reporters at run-time. It could get weird, that's all. Since the dependency isn't declared anywhere, it isn't managed by npm, so there's technically no guarantee the bundled reporters will be compatible with an installed reporters package.

Just FYI, if this wasn't clear. If you don't care, that's your call. 🙂
